File: python/techai_webutils/src/techai_webutils/clients/cache/envelope.py
"""Cache envelope for typed, version-aware caching.

Provides a generic CacheEnvelope[T] that wraps cached values with schema
version metadata. When the envelope's version doesn't match the expected
version, the cached entry is treated as a miss.
"""

from __future__ import annotations

from dataclasses import dataclass
import json
import time


@dataclass(frozen=True, slots=True)
class CacheEnvelope[T]:
    """Wraps a cached value with metadata for version-aware caching."""

    data: T
    """The cached value being wrapped."""
    cached_at: float
    """Unix timestamp (seconds) when the value was cached."""
    version: int
    """Schema version of the cached value; a mismatch on read is treated as a miss."""


def encode(value: object, version: int) -> bytes:
    """Serialize value into a versioned envelope as JSON bytes."""
    envelope = {
        "data": value,
        "cached_at": time.time(),
        "version": version,
    }
    return json.dumps(envelope).encode("utf-8")


def decode(data: bytes, version: int) -> tuple[object | None, bool]:
    """Deserialize bytes into a versioned envelope.

    Returns ``(value, True)`` on success and ``(None, False)`` on version
    mismatch. Raises on corrupt/invalid JSON (distinct from version mismatch).

    Args:
        data: JSON bytes produced by :func:`encode`.
        version: Expected schema version. Mismatch → ``(None, False)``.

    """
    envelope = json.loads(data)
    if envelope["version"] != version:
        return None, False
    return envelope["data"], True

File: go/foundation/cache/envelope.go
// Package cache provides caching utilities for the foundation layer.
package cache

import (
	"encoding/json"
	"time"
)

// Envelope wraps a cached value with metadata for version-aware caching.
// When the envelope's version doesn't match the expected version, the cached
// entry is treated as a miss — stale-schema entries are silently evicted.
//
// Serialization is JSON, deliberately, not proto/binary: T is an arbitrary type
// parameter (repository entities: Ent models, domain structs), NOT a proto.Message,
// so proto.Marshal cannot encode it. A cache of one *concrete* proto response type
// can use proto encoding; that pattern does not generalize to Envelope[T].
// The companion cursor codec (repos/repository/cursor.go) drops reflection where
// it can (a fixed 2-field shape); the generic payload here cannot, and a
// binary-header-plus-JSON-payload variant would churn the cache wire format on a
// hot read path for a small, version-bump-only gain — so JSON stays.
type Envelope[T any] struct {
	// Data is the cached value payload.
	Data T `json:"data"`

	// CachedAt records when the envelope was written, for observability and TTL reasoning.
	CachedAt time.Time `json:"cached_at"`

	// Version is the schema version the value was encoded with; a mismatch on
	// Decode is treated as a miss so stale-schema entries self-evict.
	Version int `json:"version"`
}

// Encode serializes value into a versioned envelope as JSON bytes.
func Encode[T any](value T, version int) ([]byte, error) {
	env := Envelope[T]{
		Data:     value,
		CachedAt: time.Now(),
		Version:  version,
	}
	return json.Marshal(env)
}

// Decode deserializes bytes into a versioned envelope. Returns (value, true, nil)
// on success, (zero, false, nil) on version mismatch (stale-schema miss), and
// (zero, false, err) on corrupt/invalid JSON.
func Decode[T any](data []byte, version int) (val T, ok bool, err error) {
	var env Envelope[T]
	if err := json.Unmarshal(data, &env); err != nil {
		var zero T
		return zero, false, err
	}
	if env.Version != version {
		var zero T
		return zero, false, nil
	}
	return env.Data, true, nil
}
